core/trino-main/src/main/java/io/trino/server/security/PasswordAuthenticator.java

```diff
@@ -65,6 +74,25 @@ public Identity authenticate(ContainerRequestContext request)
         }
     }
 
+    /**
+     * When the user in the basic authentication header matches the x-trino-user header, we assume that the client does
+     * not want to force the runtime user name, and only wanted to communicate the authentication user.
+     */
+    private void clearRequestUserHack(BasicAuthCredentials basicAuthCredentials, MultivaluedMap<String, String> headers)
+    {
+        String userHeader;
+        try {
+            userHeader = detectProtocol(alternateHeaderName, headers.keySet()).requestUser();
+        }
+        catch (ProtocolDetectionException ignored) {
+            // this shouldn't fail here, but ignore and it will be handled elsewhere
+            return;
+        }
+        if (basicAuthCredentials.getUser().equals(headers.getFirst(userHeader))) {
+            headers.remove(userHeader);
```

Comment:
This modifies header state which can is technically legal, but should be voided as much as possible.
X-Trino-User can be inspected by other components -- and it indeed is, otherwise you wouldn't remove it in the first place.

Instead of modifying the state (and removing factually correct information), you can add a request attribute and inspect it in the place where you want the logic to be different if authentication happened over password authenticator.

Reply:
An alternative approach would be replacing the header value with the username obtained from user mapping which might be a preferred if there are are any assumptions about existence of this header (which looks required in protocol API).